### PR TITLE
[Pubgrub] Add support for branch dependencies

### PR DESCRIPTION
**Pull request by kiliankoe**
Add support for branch-based dependencies.

### CODE REVIEW

File: Sources/PackageGraph/Pubgrub.swift [hunk 1]
```diff
@@ -187,6 +201,10 @@ extension PackageRequirement {
         switch (self, other) {
         case (.versionSet(let lhs), .versionSet(let rhs)):
             return lhs.intersection(rhs) == rhs
+        case (.unversioned, .unversioned):
+            return true
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/Pubgrub.swift]: This might not be correct since `unversioned` is always supposed to "win". Can you keep `unversioned` out of this PR? We need to add it carefully to avoid slipping in bugs that can be hard to find.

File: Sources/PackageGraph/Pubgrub.swift [hunk 2]
```diff
@@ -456,9 +481,9 @@ final class PartialSolution {
 
     /// Create a new decision assignment and add it to the partial solution's
     /// list of known assignments.
-    func decide(_ package: PackageReference, atExactVersion version: Version) {
+    func decide(_ package: PackageReference, at version: BoundVersion) {
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/Pubgrub.swift]: Can you make another PR that only contains the `BoundVersion` API changes? That will help with simplifying the diffs for the branch-based dependencies.

File: Sources/PackageGraph/Pubgrub.swift [hunk 3]
```diff
@@ -1233,41 +1258,62 @@ public final class PubgrubDependencyResolver {
...
+                // FIXME:
+                //
+                // If the selected version is the latest version, Pubgrub
+                // represents the term as having an unbounded upper range.
+                // We can't represent that here (currently), so we're
+                // pretending that it goes to the next nonexistent major
+                // version.
+                let nextMajor = Version(version.major + 1, 0, 0)
+                terms.append(Term(container.identifier, .versionSet(.range(version..<nextMajor))))
+                terms.append(Term(not: dep.identifier, dep.requirement))
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/Pubgrub.swift]: Let's just `fatalError` for now if the requirement from a versioned dependency is not another versioned requirement.

File: Sources/PackageGraph/Pubgrub.swift [hunk 4]
```diff
@@ -1233,41 +1258,62 @@ public final class PubgrubDependencyResolver {
...
+            }
+        case .revision(let revision):
+            return try container.getDependencies(at: revision).map { dep -> Incompatibility in
+                var terms: OrderedSet<Term> = []
+                terms.append(Term(container.identifier, .revision(revision)))
+                terms.append(Term(not: dep.identifier, dep.requirement))
+                return Incompatibility(terms, root: root!, cause: .dependency(package: container.identifier))
+            }
+        case .excluded:
+            fatalError()
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/Pubgrub.swift]: Can you add some explanation in the fatalError here.

File: Sources/PackageGraph/Pubgrub.swift [hunk 5]
```diff
@@ -1298,3 +1344,18 @@ public final class PubgrubDependencyResolver {
...
     }
 }
+
+extension BoundVersion {
+    fileprivate func toRequirement() -> PackageRequirement {
+        switch self {
+        case .version(let version):
+            return .versionSet(.exact(version))
+        case .excluded:
+            fatalError()
```
Comment (aciidb0mb3r) [inline Sources/PackageGraph/Pubgrub.swift]: Can you add some explanation in the fatalError here.

File: Tests/PackageGraphTests/PubgrubTests.swift [hunk 6]
```diff
@@ -917,3 +972,14 @@ extension PackageReference: ExpressibleByStringLiteral {
         self.init(identity: name, path: "")
     }
 }
+
+extension BoundVersion: Comparable {
+    public static func < (lhs: BoundVersion, rhs: BoundVersion) -> Bool {
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PubgrubTests.swift]: Can you make do without this since BoundVersion is not really Comparable?

File: Tests/PackageGraphTests/PubgrubTests.swift [hunk 7]
```diff
@@ -801,6 +834,28 @@ final class PubgrubTests: XCTestCase {
         }
     }
 
+    func testResolutionWithBranchBasedDependency() {
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PubgrubTests.swift]: Can you add some more cases (including unsatisfiable ones) of branch-based dependencies?

File: Tests/PackageGraphTests/PubgrubTests.swift [hunk 8]
```diff
@@ -358,6 +378,14 @@ final class PubgrubTests: XCTestCase {
...
         let notEmptyA = Term(not: "a", .versionSet(.empty))
         XCTAssertNil(term("a^1.0.0").difference(with: notEmptyA))
+
+        let aUnversioned = Term("a", .unversioned)
+        XCTAssertEqual(aUnversioned.intersect(with: aUnversioned), aUnversioned)
+
+        let notAUnversioned = Term(not: "a", .unversioned)
+        XCTAssertEqual(aUnversioned.intersect(with: notAUnversioned), nil)
+
+        XCTAssertEqual(notAUnversioned.intersect(with: notAUnversioned), notAUnversioned)
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PubgrubTests.swift]: We need more intersection tests for branch-based deps.

File: Tests/PackageGraphTests/PubgrubTests.swift [hunk 9]
```diff
@@ -389,6 +417,11 @@ final class PubgrubTests: XCTestCase {
         XCTAssertEqual(term("¬a^1.0.0").relation(with: "¬a^1.5.0"), .subset)
         XCTAssertEqual(term("¬a^2.0.0").relation(with: "¬a^1.0.0"), .overlap)
         XCTAssertEqual(term("¬a^1.5.0").relation(with: "¬a^1.0.0"), .overlap)
+
+        let aUnversioned = Term("a", .unversioned)
+        XCTAssertEqual(aUnversioned.relation(with: aUnversioned), .subset)
+        let notAUnversioned = Term(not: "a", .unversioned)
+        XCTAssertEqual(aUnversioned.relation(with: notAUnversioned), .disjoint)
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/PubgrubTests.swift]: We need more relation tests for branch-based deps.

**Comment by kiliankoe** [comment]
@swift-ci please smoke test

**Comment by BenchR267** [comment]
Sorry, wanted to scroll and clicked instead 🙈 Bad UX 😆

File: Tests/PackageGraphTests/XCTestManifests.swift [hunk 10]
```diff
@@ -66,6 +66,8 @@ extension PubgrubTests {
         ("testResolutionNoConflicts", testResolutionNoConflicts),
         ("testResolutionNonExistentVersion", testResolutionNonExistentVersion),
         ("testResolutionPerformingConflictResolution", testResolutionPerformingConflictResolution),
+        ("testResolutionWithOverridingBranchBasedDependency", testResolutionWithOverridingBranchBasedDependency),
+        ("testResolutionWithSimpleBranchBasedDependency", testResolutionWithSimpleBranchBasedDependency),
```
Comment (aciidb0mb3r) [inline Tests/PackageGraphTests/XCTestManifests.swift]: Can you update linux main for the new tests?

Comment (kiliankoe) [reply Tests/PackageGraphTests/XCTestManifests.swift]: 🤦‍♂️ 
Every. single. time.

**Comment by kiliankoe** [comment]
@swift-ci please smoke test

**Comment by kiliankoe** [comment]
@swift-ci test linux

**Comment by aciidb0mb3r** [comment]
Going to merge this since full Linux test passed and smoke test is a smaller subset of it.
